doc/classes/TileMap.xml

```diff
@@ -377,5 +377,14 @@
 		<constant name="TILE_ORIGIN_BOTTOM_LEFT" value="2" enum="TileOrigin">
 			Tile origin at its bottom-left corner.
 		</constant>
+		<constant name="COLLISION_VISIBILITY_MODE_ALWAYS_SHOW" value="0" enum="CollisionVisibilityMode">
+			Always show collision shapes in editor and game.
+		</constant>
```

Comment:
I'm a bit confused by these options. There doesn't seem to be a way to show collision shapes in editor, but not in game? That would seem like the most common use case to me.

Reply:
This property is mostly a switch to debug Tilemap collision in both editor and game, with a default option that allows to switch it off on all tilemaps at the same time, but it's true that a 4th option for editor only could be useful too.

Reply:
To me the expected behavior would be that it behaves like other collision shapes by default:
- Always visible in the editor
- Visible in game if "Visible Collision Shapes" is toggled

And then since it can be annoying to have tilemaps always show collision shapes in the editor, or in game with Visible Collision Shapes, there needs to be a way to turn it off.

So IMO a boolean as done initially should have been enough, but maybe its logic was not correct?

- `show_collision = true`: collision shapes visible in editor, visible in game if collision debugging is enabled, otherwise not visible
- `show_collision = false`: collision shapes not visible in editor nor in game, regardless of debug settings

Is there any other use case worth supporting via an enum?

Reply:
This behavior is what #46623 was originally doing, but it didn't allow hiding all tilemaps at once in the editor when several of them are used as layers, which could cause some problems (see https://github.com/godotengine/godot/pull/46623#issuecomment-801554658).

This PR implements the solution proposed by @groud in https://github.com/godotengine/godot/pull/46623#issuecomment-801877455, which is to have more options instead of a boolean, so by default it can be controlled using "Visible Collision Shapes" option for all tilemaps.

A gizmo could be a less confusing alternative to enable/disable collision shapes in the editor, but I don't know if it's possible to have one that would affect only the collision from tilemaps, it seems you can have only one gizmo per node type.

Reply:
The "Visible Collision Shapes" option is used to configure what is visible at *runtime*, not in the editor. Collision shapes are always visible in the editor, regardless of whether that option is enabled or not.

So IMO it makes things more confusing if this option starts being used to show/hide tilemap collision shapes in the editor, while still leaving non-tilemap collision shapes always visible.

If "Visible Collision Shapes" is turned off, tilemap collision shapes should *never* be visible at runtime (but the current documentation suggests for the enum values otherwise).

https://github.com/godotengine/godot/pull/46623#issuecomment-801554658 is a misunderstanding of what "Visible Collision Shapes" is intended for. For this user's use case, they should just disable collision shapes visibility for all the tilemaps they don't want to see.

OR we simply switch the boolean to false by default, and let users enable it for the tilemaps where they do want to see visible collision shapes.

Reply:
I agree with akien. `show_collision` should only effect in-editor. This way things are more simple.

Reply:
> I agree with akien. `show_collision` should only effect in-editor. This way things are more simple.

To clarify, `show_collision = true` + "Visible Collision Shapes" turned ON should always show tilemap collision shapes in-game.
Basically `show_collision` should control "is this TileMap used in the collision shapes debug visualization?"

And that collision shapes debug visualization, for all collision shapes + enabled TileMaps, can happen in two ways:
- Always on in the editor
- In-game depending on "Visible Collision Shapes"

Reply:
> To clarify, show_collision = true + "Visible Collision Shapes" turned ON should also show tilemap collision shapes in-game.

What I am saying is to use `show_collision` to control visibility in-editor and use "Visible Collision Shapes" to control visibility in-game. Keeping both separate. 

>  Basically show_collision should control "is this TileMap used in the collision shapes debug visualization?"

This is what actually #46623 does.

Reply:
> This is what actually #46623 does.

Yeah that's why I think it was fine as is. If it's a bother to too many users, I'd just change the default to be disabled by default, and users can enable it for the tilemaps where they want to debug collisions.

But maybe it's just a handful of users using 10 tilemaps and wanting to save 10 clicks to disable the collisions, while most users using a single tilemap would benefit from it being on by default (if only for discoverability). So I wouldn't make the behavior a lot more complex and different from other types of collision shapes too fast.

Reply:
I agree again. IMO its special use case(https://github.com/godotengine/godot/pull/46623#issuecomment-801661554)

Reply:
@akien-mga Old pr has [this](https://github.com/godotengine/godot/pull/47204#issuecomment-803526151) problem, let me know if we are not merging this one, I need to update the old ones.

Reply:
My idea for this would be to make the "Show debug collision shape" settings smarter. And change it to a 3-options too:
- Hide,
- Show in editor,
- Show in editor and in game,

The question is what to do with the shapes. If the idea is to override this setting, we might simply have instead those 4 options per shape:
- Use global settings,
- Hide,
- Show in editor,
- Show in editor and in game.

I believe this is kind of the most expected way to set all of this, as those 3 options (plus one to "inherit" settings) are basically 3 levels of debug. From the "In don't want to debug physics" to the "I have issues with physics, show me shapes in game".

Reply:
> OR we simply switch the boolean to false by default, and let users enable it for the tilemaps where they do want to see visible collision shapes.

Maybe this is actually the easiest option. This way it's disabled by default in both editor and game (which solves https://github.com/godotengine/godot/pull/46623#issuecomment-801554658), and then users can set things individually:
-`show_collision=false`: always disabled in both game and editor
-`show_collision=true`: always enabled in editor, controlled by "Visible Collision Shapes" in game

So the only thing that would not be possible for now is to switch collisions on and off for all tilemaps at once in the editor, but that could be handled later with an improved gizmo system.

Reply:
I opened #47382 to switch it to opt-in.